lambeq/training/nelder_mead_optimizer.py

```diff
@@ -0,0 +1,394 @@
+# Copyright 2021-2023 Cambridge Quantum Computing Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+NelderMeadOptimizer
+=============
+Module implementing the Nelder-Mead Optimization Algorithm.
+
+"""
+from __future__ import annotations
+
+from collections.abc import Callable, Iterable, Mapping
+from typing import Any
+import warnings
+
+import numpy as np
+from numpy.typing import ArrayLike
+
+from lambeq.core.utils import flatten
+from lambeq.training.optimizer import Optimizer
+from lambeq.training.quantum_model import QuantumModel
+
+
+class NelderMeadOptimizer(Optimizer):
+    """Nelder Mead Optimizer
+
+    The Nelder-Mead optimizer is an algorithm used for unconstrained
+    optimization in multidimensional spaces. Unlike some other
+    optimization methods, it does not take into account any bounds or
+    constraints on the variables. The algorithm is based on the Simplex
+    method and is particularly useful when the derivatives
+    (first and second)  of the objective function are unknown or
+    unreliable.
+
+    Although the Nelder-Mead algorithm is generally robust and widely
+    applicable, it has some limitations. In cases where the derivatives
+    can be accurately computed, alternative algorithms that utilize
+    this derivative information may offer better performance. These
+    methods are often preferred due to their ability to handle a wider
+    range of scenarios and their tendency to converge to more optimal
+    solutions. It is worth noting that the Nelder-Mead technique is a
+    heuristic search approach, which means that it can sometimes
+    converge to non-stationary points or suboptimal solutions.
+
+    This implementation is based heavily scipy's optimize.minimize. See
+    https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.minimize.html
+
+    """
+
+    model: QuantumModel
+
+    def __init__(
+        self,
+        model: QuantumModel,
+        hyperparams: dict[str, float],
+        loss_fn: Callable[[Any, Any], float],
+        bounds: ArrayLike | None = None,
+    ) -> None:
+        """Initialise the Nelder-Mead optimizer.
+
+        The hyperparameters may contain the following key value pairs::
+
+        hyperparams = {
+            'adpative': Adjust the algorithm's parameters based on the
```

Comment:
I really like the docstrings you've written in general, they're very informative. However, I don't really like this little section. I think the explanations are too long to be mixed with the syntax like this. Perhaps it would look better if you separated it out?

Also there's a typo on this line.

Reply:
I have tried to make it look nicer. still it it does not feel good, a workaround would be to show only the keys of the dictionary and point the user to Scipy's docs.

Reply:
Shall I consider this resolved?